packages/angular_devkit/build_angular/src/karma/index.ts

```diff
@@ -95,6 +112,31 @@ export function execute(
         }
       }
 
+      // prepend special webpack loader that will transform test.ts
+      if (webpackConfig && webpackConfig.module && options.spec) {
+        const sourceRoot = workspace.getProject(projectName).sourceRoot || '';
```

Comment:
Why is `sourceRoot` used here?
The current behavior with the `require.context` call is that the files are relative to the `test.ts` (`mainFilePath` in this case).
However, I think it would be more intuitive if the option's base directory was the workspace root as this would be consist with all other options.  A project might also not have the concept of a source root (its currently really an extension option for the browser builder).
This would also eliminate the need for all the extra code to find the source root.

Reply:
Well, option's base directory is workspace root - path relative to source root is an alternative that is exposed, but mainly needed internally. `sourceRoot` is simply subtracted from what is passed as `spec`, so that typescript would be able to find these files in `test.ts` (main) as path relative to it is needed (see [findSpecs](https://github.com/angular/angular-cli/pull/13423/files#diff-dccf1f76f21004c8dc59dd26881c6816R13)).

Any suggestions on how to refactor this to not rely on sourceRoot?

Reply:
The `glob.sync` call's `cwd` option can be the workspace root and the `absolute` option would also be needed.  From there, get the absolute path of the directory name of the main file; and then loop and convert the test file list from the glob call to relative paths from the main file.

Reply:
I've ended up with a bit different solution - sourceRoot is no longer used and I've reverted the changes. I hope it's ok now.